File: sdk/javascript/examples/complete-workflow.ts
/**
 * Example: Complete Workflow
 *
 * Demonstrates the full Slaver lifecycle:
 *   1. Register as a slaver agent
 *   2. Claim an available task
 *   3. Maintain heartbeats while working
 *   4. Submit a pull request
 *   5. Notify master and clean up
 *
 * Run: npx ts-node sdk/javascript/examples/complete-workflow.ts
 */

import { EketClient, sleep } from '../src/index.js';

const SERVER_URL = process.env.EKET_SERVER_URL ?? 'http://localhost:8080';

async function main() {
  const client = new EketClient({ serverUrl: SERVER_URL });

  // ── Step 1: Register ──────────────────────────────────────────────────────
  console.log('Step 1: Registering slaver agent...');
  const { instance_id } = await client.registerAgent({
    agent_type: 'claude_code',
    agent_version: '1.0.0',
    role: 'slaver',
    specialty: 'fullstack',
    capabilities: ['typescript', 'react', 'nodejs', 'testing'],
    metadata: { user: 'developer', machine: process.env.HOSTNAME ?? 'local' },
  });
  console.log(`✓ Registered: ${instance_id}`);

  // ── Step 2: Find and claim a task ─────────────────────────────────────────
  console.log('\nStep 2: Looking for available tasks...');
  const available = await client.listTasks({ status: 'ready' });
  console.log(`  Found ${available.length} task(s)`);

  if (available.length === 0) {
    console.log('ℹ No ready tasks — exiting.');
    await client.deregisterAgent(instance_id);
    return;
  }

  const selected = available[0];
  console.log(`  Selecting: [${selected.id}] ${selected.title}`);

  console.log('\nStep 3: Claiming task...');
  const claimed = await client.claimTask(selected.id, instance_id);
  console.log(`✓ Claimed: ${claimed.id} (status=${claimed.status})`);

  // ── Step 3: Work with heartbeats ─────────────────────────────────────────
  console.log('\nStep 4: Working on task (with heartbeats)...');

  const workPhases: Array<{ label: string; progress: number }> = [
    { label: 'Analysing requirements', progress: 0.1 },
    { label: 'Scaffolding code',       progress: 0.3 },
    { label: 'Core implementation',    progress: 0.6 },
    { label: 'Writing tests',          progress: 0.85 },
    { label: 'Final review',           progress: 1.0 },
  ];

  for (const phase of workPhases) {
    console.log(`  [${Math.round(phase.progress * 100)}%] ${phase.label}`);

    // Send heartbeat so the server knows we're alive
    const hb = await client.sendHeartbeat(instance_id, {
      status: 'busy',
      current_task: claimed.id,
      progress: phase.progress,
    });

    if (hb.messages.length > 0) {
      console.log(`  📧 ${hb.messages.length} pending message(s) from server`);
    }

    // Update task progress in the tracker
    await client.updateTask(claimed.id, {
      progress: phase.progress,
      notes: phase.label,
    });

    await sleep(500); // simulate work
  }

  // ── Step 4: Mark ready for review ────────────────────────────────────────
  console.log('\nStep 5: Marking task as "review"...');
  await client.updateTask(claimed.id, { status: 'review', progress: 1.0 });
  console.log('✓ Task status → review');

  // ── Step 5: Submit PR ─────────────────────────────────────────────────────
  console.log('\nStep 6: Submitting pull request...');
  const prId = await client.submitPR({
    instance_id,
    task_id: claimed.id,
    branch: `feature/${claimed.id.toLowerCase()}-implementation`,
    description: `Implements ${claimed.title}.\n\n- All acceptance criteria met\n- Tests passing`,
    test_status: 'passed',
  });
  console.log(`✓ PR submitted: ${prId}`);

  // ── Step 6: Notify master ─────────────────────────────────────────────────
  console.log('\nStep 7: Notifying master for review...');
  await client.sendMessage({
    from: instance_id,
    to: 'master',
    type: 'pr_review_request',
    priority: 'normal',
    payload: {
      task_id: claimed.id,
      branch: `feature/${claimed.id.toLowerCase()}-implementation`,
      pr_id: prId,
      description: `Please review: ${claimed.title}`,
    },
  });
  console.log('✓ Master notified');

  // ── Cleanup ───────────────────────────────────────────────────────────────
  console.log('\nStep 8: Cleaning up...');
  await client.sendHeartbeat(instance_id, { status: 'idle' });
  await client.deregisterAgent(instance_id);
  await client.shutdown();
  console.log('✓ Agent deregistered');

  console.log('\n✅ Complete workflow finished!');
}

main().catch((err: unknown) => {
  const message = err instanceof Error ? err.message : String(err);
  console.error('❌ Fatal error:', message);
  process.exit(1);
});

File: sdk/python/examples/complete_workflow.py
#!/usr/bin/env python3
"""
Example: Complete Workflow

This example demonstrates a complete workflow from registration to PR merge.
"""

import time
from eket_sdk import (
    EketClient,
    AgentType,
    AgentRole,
    AgentSpecialty,
    AgentStatus,
    TaskStatus,
    MessageType,
)
from eket_sdk.models import TestStatus


def main():
    # Initialize client
    print("🚀 Starting EKET Complete Workflow Example\n")
    client = EketClient(server_url="http://localhost:8080")

    # Step 1: Register agent
    print("📝 Step 1: Register Agent")
    agent = client.register_agent(
        agent_type=AgentType.CUSTOM,
        role=AgentRole.SLAVER,
        specialty=AgentSpecialty.FULLSTACK,
        capabilities=["python", "react", "typescript", "postgresql"],
        metadata={"user": "demo_user", "machine": "demo-workstation"},
    )
    print(f"   ✅ Registered as {agent.instance_id}\n")

    # Step 2: Send initial heartbeat
    print("💓 Step 2: Send Initial Heartbeat")
    heartbeat_response = client.send_heartbeat(status=AgentStatus.IDLE)
    print(f"   ✅ Heartbeat acknowledged at {heartbeat_response.get('server_time')}\n")

    # Step 3: List available tasks
    print("📋 Step 3: List Available Tasks")
    tasks = client.list_tasks(status=TaskStatus.READY)
    if not tasks:
        print("   ⚠️  No tasks available")
        client.deregister_agent()
        return

    print(f"   Found {len(tasks)} available task(s):")
    for task in tasks[:3]:
        print(
            f"   - {task.id}: {task.title} [{task.priority.value}] ({task.estimate or 'no estimate'})"
        )
    print()

    # Step 4: Claim a task
    selected_task = tasks[0]
    print(f"🎯 Step 4: Claim Task {selected_task.id}")
    claimed_task = client.claim_task(selected_task.id)
    print(f"   ✅ Task claimed successfully")
    print(f"   Status: {claimed_task.status.value}")
    print(f"   Assigned to: {claimed_task.assigned_to}\n")

    # Step 5: Work on task with progress updates
    print(f"💼 Step 5: Work on Task (simulated)")
    for progress in [0.25, 0.5, 0.75, 1.0]:
        time.sleep(1)  # Simulate work

        # Send heartbeat
        client.send_heartbeat(
            status=AgentStatus.BUSY, current_task=claimed_task.id, progress=progress
        )

        # Update task
        client.update_task(claimed_task.id, progress=progress)
        print(f"   Progress: {int(progress * 100)}%")

    print()

    # Step 6: Update task to review status
    print(f"✅ Step 6: Mark Task as Complete")
    completed_task = client.update_task(
        claimed_task.id,
        status=TaskStatus.REVIEW,
        progress=1.0,
        notes="Implementation completed, ready for review",
    )
    print(f"   Task status: {completed_task.status.value}\n")

    # Step 7: Submit PR
    print(f"📤 Step 7: Submit Pull Request")
    branch = f"feature/{claimed_task.id}-implementation"
    pr_id = client.submit_pr(
        instance_id=agent.instance_id,
        task_id=claimed_task.id,
        branch=branch,
        description=f"""
## Summary
Completed implementation for {claimed_task.title}

## Changes
- Implemented core functionality
- Added unit tests
- Updated documentation

## Test Results
- All tests passing
- Coverage: 85%
        """,
        test_status=TestStatus.PASSED,
    )
    print(f"   ✅ PR submitted: {pr_id}")
    print(f"   Branch: {branch}\n")

    # Step 8: Send PR review request to master
    print(f"📨 Step 8: Send Review Request to Master")
    masters = client.list_agents(role=AgentRole.MASTER)
    if masters:
        master_id = masters[0].instance_id
        message_id = client.send_message(
            from_id=agent.instance_id,
            to_id=master_id,
            msg_type=MessageType.PR_REVIEW_REQUEST,
            payload={
                "task_id": claimed_task.id,
                "branch": branch,
                "pr_id": pr_id,
                "description": completed_task.title,
                "test_status": "passed",
            },
        )
        print(f"   ✅ Review request sent to {master_id}")
        print(f"   Message ID: {message_id}\n")
    else:
        print("   ⚠️  No master agents available\n")

    # Step 9: Check for messages
    print(f"📬 Step 9: Check for Messages")
    messages = client.get_messages()
    if messages:
        print(f"   Received {len(messages)} message(s):")
        for msg in messages:
            print(f"   - {msg.type.value} from {msg.from_id}")
    else:
        print("   No new messages\n")

    # Step 10: Final heartbeat and deregister
    print(f"👋 Step 10: Cleanup")
    client.send_heartbeat(status=AgentStatus.IDLE)
    client.deregister_agent()
    print(f"   ✅ Agent deregistered\n")

    print("✨ Workflow completed successfully!")


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n\n⏹️  Interrupted by user")
    except Exception as e:
        print(f"\n❌ Error: {e}")
        import traceback

        traceback.print_exc()
